### PR TITLE
[Feat/aut510] - SmartPeak support for isotopic labeling experiments

### PR DESCRIPTION
**Pull request by ghost**
**Description**
Wrapping the OpenMS class and methods for isotopic labelling experiments in SmartPeak so that all methods can be accessed through the SmartPeak workflow engine. The following RawDataProcessor structs are added:

- `CalculateMDVs`
- `IsotopicCorrections`
- `CalculateIsotopicPurities`
- `CalculateMDVAccuracies`

**RawDataProcessor Parameters for IsotopeLabelingMDVs**
The following parameters can be saved in the `RawDataProcessor_params.csv` file:
- `CalculateMDVs` 
  - `mass_intensity_type` :` norm_sum` or `norm_max`
  - `feature_name` : `peak_apex_int` or `intensity` 
- `IsotopicCorrections`
  - `correction_matrix_agent` : `TBDMS`
- `CalculateIsotopicPurities`
  -  `isotopic_purity_name` : e.g. `1_2-13C_glucose_experiment` -> can be extracted as `MetaValue`.
  - `isotopic_purity_values` : list of values e.g. `"[0.5, 0.7, 98.8, 0.0, 0.0, 0.0]"`
- `CalculateMDVAccuracies`
  - `fragment_isotopomer_measured` : list of values e.g. `"[0.627, 0.253, 0.096, 0.02, 0.004, 0.001]"` where `description` is the `ProteinName` e.g. `accoa`

~~**Note**
This PR is using a forked OpenMS with support for isotopic labeling experiments being developed in [OpenMS:feat/aut509](https://github.com/ahmedskhalil/OpenMS/tree/feat/aut509), the CI relies on this build to execute the pipelines for all platforms. Once this branch is merged into `OpenMS/develop`, the CI shall be changed to point to `OpenMS/develop`.~~ -> Updated.

**Future Improvements**
- Adding Copyright notice to source and header files, this could also be done in this PR.

### CODE REVIEW

File: external/plog.cmake [hunk 1]
```diff
@@ -3,7 +3,7 @@ include(${CMAKE_ROOT}/Modules/ExternalProject.cmake)
 set_property (DIRECTORY PROPERTY EP_BASE Dependencies)
 
 set (DEPENDENCIES)
-set (PLOG_VERSION 5f69a631086bf038e2c0feee4319a485c34fa863)
+set (PLOG_VERSION 1.1.4)
```
Comment (dmccloskey) [inline external/plog.cmake:6]: Is there a reason for this change?

File: src/smartpeak/include/SmartPeak/core/RawDataProcessor.h [hunk 2]
```diff
@@ -561,4 +563,64 @@ namespace SmartPeak
...
+                 RawDataHandler& rawDataHandler_IO,
+                 const std::map<std::string, std::vector<std::map<std::string, std::string>>>& params_I,
+                 const Filenames& filenames
+                 ) const override;
+  };
+  
+  struct IsotopicCorrections : RawDataProcessor
+  {
+    int getID() const override { return 0; }
+    std::string getName() const override { return "ISOTOPIC_CORRECTIONS"; }
```
Comment (dmccloskey) [inline src/smartpeak/include/SmartPeak/core/RawDataProcessor.h:627]: I am kind of partial to "CORRECT_MDVS" following your description on the line below.

File: src/smartpeak/source/core/RawDataProcessor.cpp [hunk 3]
```diff
@@ -1844,4 +1844,206 @@ namespace SmartPeak
...
+          mass_intensity_type = param.find("value")->second;
+        }
+        else if (param.find("name")->second == "feature_name")
+        {
+          feature_name = param.find("value")->second;
+        }
+      }
+      
+      isotopelabelingmdvs.calculateMDVs(rawDataHandler_IO.getFeatureMap(), normalized_featureMap, mass_intensity_type, feature_name);
+      rawDataHandler_IO.setFeatureMap(normalized_featureMap);
```
Comment (dmccloskey) [inline src/smartpeak/source/core/RawDataProcessor.cpp:1958]: Don't forget the call to `rawDataHandler_IO.updateFeatureMapHistory();` to update the Feature map history with the new feature data.

Comment (dmccloskey) [reply src/smartpeak/source/core/RawDataProcessor.cpp:1958]: Also in the rest of the new methods.

File: src/smartpeak/source/core/RawDataProcessor.cpp [hunk 4]
```diff
@@ -1844,4 +1844,206 @@ namespace SmartPeak
...
+
+    try {
+      OpenMS::FeatureMap normalized_featureMap;
+      std::vector<std::map<std::string, std::string>> CalculateIsotopicPurities_params = params_I.find("CalculateIsotopicPurities")->second;
+      
+      std::string experiment_data_s, isotopic_purity_name;
+      std::vector<double> experiment_data;
+      for(auto& param : CalculateIsotopicPurities_params)
+      {
+        if (!param.find("comment_")->second.empty())
```
Comment (dmccloskey) [inline src/smartpeak/source/core/RawDataProcessor.cpp]: What is the reason for using the `comment_` column?   It is actually not supposed to be used for the data processing, but only for the user when directly annotating the parameters file.

File: src/tests/class_tests/smartpeak/source/RawDataProcessor_test.cpp [hunk 5]
```diff
@@ -2345,4 +2345,309 @@ BOOST_AUTO_TEST_CASE(emg_processor)
...
+{
+  // Pre-requisites: load the parameters and associated raw data
+  map<string, vector<map<string, string>>> params_1;
+  map<string, vector<map<string, string>>> params_2;
+  load_data(params_1, params_2);
+  RawDataHandler rawDataHandler;
+
+  Filenames filenames;
+  filenames.featureXML_i = SMARTPEAK_GET_TEST_DATA_PATH("GCMS_SIM.featureXML");
+//
```
Comment (dmccloskey) [inline src/tests/class_tests/smartpeak/source/RawDataProcessor_test.cpp]: Are these commented lines needed here?

**Review by dmccloskey — COMMENTED** [review]
I have added my initial comments.  This is looking nice.  Let's talk more tomorrow at our meeting.

File: external/plog.cmake (hunk 1 above)
Comment (ghost) [reply external/plog.cmake:6]: Just avoiding checking out by commit hashes and opting for a stable release instead, I could revert it to `5f69a631086bf038e2c0feee4319a485c34fa863` if that's preferred.

File: src/smartpeak/include/SmartPeak/core/RawDataProcessor.h (hunk 2 above)
Comment (ghost) [reply src/smartpeak/include/SmartPeak/core/RawDataProcessor.h:627]: Thanks for the hint.

File: src/smartpeak/source/core/RawDataProcessor.cpp (hunk 3 above)
Comment (ghost) [reply src/smartpeak/source/core/RawDataProcessor.cpp:1958]: Done.

File: src/smartpeak/source/core/RawDataProcessor.cpp (hunk 4 above)
Comment (ghost) [reply src/smartpeak/source/core/RawDataProcessor.cpp]: The `calculateMDVAccuracies(..)` takes in a 2 vectors of doubles for the measured and theoretical data, could we embed these data in other `csv` files ?

File: src/tests/class_tests/smartpeak/source/RawDataProcessor_test.cpp (hunk 5 above)
Comment (ghost) [reply src/tests/class_tests/smartpeak/source/RawDataProcessor_test.cpp]: No longer needed, it was an unfinished attempt to load the features from the `featureXML` file.

File: README.rst [hunk 6]
```diff
@@ -79,8 +79,13 @@ Build OpenMS following the OpenMS wiki instructions. Example cmake command on wi
 STEP 3: Build SmartPeak dependencies
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 Download the latest SDL2 libraries. Add the SDL2 folder to the path environmental variable. Compile using cmake and build for "external projects"
+<<<<<<< HEAD
```
Comment (dmccloskey) [inline README.rst]: It looks like there are some unresolved merge conflicts?

File: src/smartpeak/include/SmartPeak/core/SampleGroupHandler.h [hunk 7]
```diff
@@ -1,4 +1,25 @@
...
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Douglas McCloskey $
+// $Authors: Douglas McCloskey, Pasquale Domenico Colaianni $
```
Comment (dmccloskey) [inline src/smartpeak/include/SmartPeak/core/SampleGroupHandler.h]: Please remove PDC on this one as he did not contribute to writing it.

File: src/smartpeak/source/core/SampleGroupHandler.cpp [hunk 8]
```diff
@@ -1,4 +1,25 @@
...
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Douglas McCloskey $
+// $Authors: Douglas McCloskey, Pasquale Domenico Colaianni $
```
Comment (dmccloskey) [inline src/smartpeak/source/core/SampleGroupHandler.cpp]: Please remove PDC on this one as he did not contribute to writing it.

File: src/smartpeak/source/core/SampleGroupProcessor.cpp [hunk 9]
```diff
@@ -1,4 +1,25 @@
...
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Douglas McCloskey $
+// $Authors: Douglas McCloskey, Pasquale Domenico Colaianni $
```
Comment (dmccloskey) [inline src/smartpeak/source/core/SampleGroupProcessor.cpp]: Please remove PDC on this one as he did not contribute to writing it.

File: src/smartpeak/source/core/RawDataHandler.cpp [hunk 10]
```diff
@@ -1,4 +1,25 @@
...
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Douglas McCloskey $
+// $Authors: Douglas McCloskey $
```
Comment (dmccloskey) [inline src/smartpeak/source/core/RawDataHandler.cpp]: Please add PDC, SK, and yourself to the Authors

File: src/smartpeak/include/SmartPeak/core/RawDataProcessor.h [hunk 11]
```diff
@@ -1,4 +1,25 @@
...
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Douglas McCloskey $
+// $Authors: Douglas McCloskey, Pasquale Domenico Colaianni $
```
Comment (dmccloskey) [inline src/smartpeak/include/SmartPeak/core/RawDataProcessor.h]: Please add SK and yourself to the Authors

File: src/smartpeak/source/core/RawDataProcessor.cpp [hunk 12]
```diff
@@ -2050,40 +2091,43 @@ namespace SmartPeak
...
+              }
+            }
           }
         }
       }
       
-      
       isotopelabelingmdvs.calculateMDVAccuracies(rawDataHandler_IO.getFeatureMap(), featureMap_with_accuracy_info,
-                                                 fragment_isotopomer_measured, fragment_isotopomer_theoretical);
+                                                 fragment_isotopomer_measured, fragment_isotopomer_theoretical_formula);
```
Comment (dmccloskey) [inline src/smartpeak/source/core/RawDataProcessor.cpp]: Is this correct?  I am thinking that this should be a vector as there will be multiple sum formulas that match each of the different transitions in the TraML file (i.e., the getTargetedExperiment() object above).

File: src/smartpeak/source/core/RawDataProcessor.cpp [hunk 13]
```diff
@@ -2003,22 +2043,23 @@ namespace SmartPeak
...
+          for (std::sregex_iterator it = values_begin; it != values_end; ++it)
+            experiment_data.push_back(std::stod(it->str()));
         }
-        if (param.find("name")->second == "isotopic_purity_name")
+        if (param.find("name")->second == "isotopic_purity_name" && !param.find("value")->second.empty())
         {
           isotopic_purity_name = param.find("value")->second;
         }
       }
       isotopelabelingmdvs.calculateIsotopicPurities(rawDataHandler_IO.getFeatureMap(), normalized_featureMap, experiment_data, isotopic_purity_name);
```
Comment (dmccloskey) [inline src/smartpeak/source/core/RawDataProcessor.cpp]: How does this method handle multiple isotopic purities?

**Review by dmccloskey — CHANGES_REQUESTED** [review]
Thanks for updating the Copyright sections.  That has been needed for sometime.

Most of the changes look great, but we should discuss how `calculateIsotopicPurities` and `calculateMDVAccuracies` are able to work with multiple components and SumFormulas respectively.

File: README.rst (hunk 6 above)
Comment (ghost) [reply README.rst]: Updated.

File: src/smartpeak/source/core/RawDataProcessor.cpp (hunk 12 above)
Comment (ghost) [reply src/smartpeak/source/core/RawDataProcessor.cpp]: Yes, I have updated the paramter list to take in `feature_name` and a map of `proteinName`s to `SumFormula`s harvested from `OpenMSFile_traML_1.csv`.

File: src/smartpeak/source/core/RawDataProcessor.cpp (hunk 13 above)
Comment (ghost) [reply src/smartpeak/source/core/RawDataProcessor.cpp]: I have modified it to handle multiple `experiment_data` as a vector of vectors and multiple `isotopic_purity_names` as vector of strings.

As for `experiment_data` there is a restriction that all sub-vectors must be of equal sizes.

File: .circleci/config.yml [hunk 14]
```diff
@@ -33,8 +33,8 @@ jobs:
           pip3 install exhale
       - run: |
           cd ~ &&
-          # git clone --branch develop --depth 1 https://github.com/OpenMS/OpenMS.git &&
-          git clone --branch feat/aut509 --depth 1 https://github.com/ahmedskhalil/OpenMS.git &&
+          git clone --branch develop --depth 1 https://github.com/OpenMS/OpenMS.git &&
+          # git clone --branch feat/aut509 --depth 1 https://github.com/ahmedskhalil/OpenMS.git &&
```
Comment (dmccloskey) [inline .circleci/config.yml]: Can this commented line be removed?  Same with the ones below.

**Review by bertrandboudaud — APPROVED** [review]
From my perspective, nothing to comment :) it looks good

File: src/tests/class_tests/smartpeak/source/Utilities_test.cpp [hunk 15]
```diff
@@ -388,10 +409,10 @@ BOOST_AUTO_TEST_CASE(getPathnameContent)
   const std::array<std::vector<std::string>, 4> c = Utilities::getPathnameContent(pathname);
 
   // number of items in the pathname, taking .gitignore into account
-  BOOST_CHECK_EQUAL(c[0].size(), 45);
```
Comment (bertrandboudaud) [inline src/tests/class_tests/smartpeak/source/Utilities_test.cpp:391]: This is a little bit boring to have to change that test every time we add files in the test directory.

**Review by dmccloskey — CHANGES_REQUESTED** [review]
Can you double check that the interfaces to OpenMS are updated?  I can see below for example:

`isotopelabelingmdvs.calculateMDVs(rawDataHandler_IO.getFeatureMap(), normalized_featureMap, mass_intensity_type, feature_name);`

That we are still calling the methods feature by feature when instead we would like apply the method to the entire `FeatureMap`.  Please let me know if you want to jump on a quick Zoom call to go over anything.

File: src/tests/class_tests/smartpeak/source/Utilities_test.cpp (hunk 15 above)
Comment (ghost) [reply src/tests/class_tests/smartpeak/source/Utilities_test.cpp:391]: we can remove it in this PR, much better would be a test that includes all related IO file operations including assertions if directory contents, etc.

**Comment by ghost** [comment]
> Can you double check that the interfaces to OpenMS are updated? I can see below for example:
> 
> `isotopelabelingmdvs.calculateMDVs(rawDataHandler_IO.getFeatureMap(), normalized_featureMap, mass_intensity_type, feature_name);`
> 
> That we are still calling the methods feature by feature when instead we would like apply the method to the entire `FeatureMap`. Please let me know if you want to jump on a quick Zoom call to go over anything.

[IsotopeLabelingMDVs::calculateMDVs](https://github.com/OpenMS/OpenMS/blob/9d2951760e313ad1a5d20ad21e9fe679c18738d5/src/openms/source/ANALYSIS/QUANTITATION/IsotopeLabelingMDVs.cpp#L325) can only accept `FeatureMap` and returns and updated `FeatureMap`, internally it's calling `IsotopeLabelingMDVs::calculateMDV` as you know, as shown in [calculateMDVs.process(rawDataHandler, params_1, filenames);](https://github.com/AutoFlowResearch/SmartPeak/blob/64a9ed4fc8470f5e6750ca72915df7bdeae6cef9/src/tests/class_tests/smartpeak/source/RawDataProcessor_test.cpp#L2555), is there something that seems to be missing?

File: src/tests/class_tests/smartpeak/source/Utilities_test.cpp (hunk 15 above)
Comment (bertrandboudaud) [reply src/tests/class_tests/smartpeak/source/Utilities_test.cpp:391]: I will create a task, it's better to focus on your task as defined for now, it's already a lot.

**Review by dmccloskey — APPROVED** [review]
Again, awesome work here!

**Pull request merged**